File: compiler/forget/src/HIR/MergeConsecutiveBlocks.ts
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import invariant from "invariant";
import {
  BlockId,
  Effect,
  GeneratedSource,
  HIRFunction,
  Instruction,
  InstructionKind,
} from "./HIR";

/**
 * Merges sequences of blocks that will always execute consecutively —
 * ie where the predecessor always transfers control to the successor
 * (ie ends in a goto) and where the predecessor is the only predecessor
 * for that successor (ie, there is no other way to reach the successor).
 *
 * Note that this pass leaves value/loop blocks alone because they cannot
 * be merged without breaking the structure of the high-level terminals
 * that reference them.
 *
 * TODO @josephsavona make value blocks explicit (eg a `kind` on Block).
 */
export function mergeConsecutiveBlocks(fn: HIRFunction): void {
  const merged = new MergedBlocks();
  for (const [, block] of fn.body.blocks) {
    // Can only merge blocks with a single predecessor, can't merge
    // value blocks
    if (block.kind !== "block" || block.preds.size !== 1) {
      continue;
    }
    const originalPredecessorId = Array.from(block.preds)[0]!;
    const predecessorId = merged.get(originalPredecessorId);
    const predecessor = fn.body.blocks.get(predecessorId);
    invariant(
      predecessor !== undefined,
      "Expected predecessor %s to exist",
      predecessorId
    );
    if (predecessor.terminal.kind !== "goto") {
      // The predecessor is not guaranteed to transfer control to this block,
      // they aren't consecutive.
      continue;
    }

    // Replace phis in the merged block with canonical assignments to the single operand value
    for (const phi of block.phis) {
      invariant(
        phi.operands.size === 1,
        "Found a block with a single predecessor but where a phi has multiple (%s) operands",
        phi.operands.size
      );
      const operand = Array.from(phi.operands.values())[0]!;
      const instr: Instruction = {
        id: predecessor.terminal.id,
        lvalue: {
          place: {
            kind: "Identifier",
            identifier: phi.id,
            effect: Effect.Mutate,
            loc: GeneratedSource,
          },
          kind: InstructionKind.Const,
        },
        value: {
          kind: "LoadLocal",
          place: {
            kind: "Identifier",
            identifier: operand,
            effect: Effect.Read,
            loc: GeneratedSource,
          },
          loc: GeneratedSource,
        },
        loc: GeneratedSource,
      };
      predecessor.instructions.push(instr);
    }

    predecessor.instructions.push(...block.instructions);
    predecessor.terminal = block.terminal;
    merged.merge(block.id, predecessorId);
    fn.body.blocks.delete(block.id);
  }
}

class MergedBlocks {
  #map: Map<BlockId, BlockId> = new Map();

  /**
   * Record that @param block was merged into @param into.
   */
  merge(block: BlockId, into: BlockId): void {
    const target = this.get(into);
    this.#map.set(block, target);
  }

  /**
   * Get the id of the block that @param block has been merged into.
   * This is transitive, in the case that eg @param block was merged
   * into a block which later merged into another block.
   */
  get(block: BlockId): BlockId {
    let current = block;
    while (this.#map.has(current)) {
      current = this.#map.get(current) ?? current;
    }
    return current;
  }
}

File: compiler/crates/react_hir/src/merge_consecutive_blocks.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::HashMap;

use react_diagnostics::{invariant, Diagnostic};
use thiserror::Error;

use crate::{
    mark_instruction_ids, mark_predecessors, BasicBlock, BlockId, BlockKind, BlockRewriter,
    BlockRewriterAction, Environment, Function, IdentifierOperand, InstrIx, Instruction,
    InstructionKind, InstructionValue, LValue, LoadLocal, StoreLocal, TerminalValue,
};

/// Merges sequences of blocks that will always execute consecutively —
/// ie where the predecessor always transfers control to the successor
/// (ends in a goto) and where the predecessor is the only predecessor
/// for that successor (ie, there is no other way to reach the successor).
///
/// Note that this pass leaves value/loop blocks alone because they cannot
/// be merged without breaking the structure of the high-level terminals
/// that reference them.
pub fn merge_consecutive_blocks(env: &Environment, fun: &mut Function) -> Result<(), Diagnostic> {
    let mut merged = MergedBlocks::default();
    let blocks = &mut fun.body.blocks;
    let instructions = &mut fun.body.instructions;
    let mut rewriter = BlockRewriter::new(blocks, fun.body.entry);
    let mut has_changes = false;

    rewriter.try_each_block(|mut block, rewriter| {
        let block_id = block.id;
        // Visit instructions to merge blocks within function expressions
        for instr_ix in &block.instructions {
            let instr = &mut instructions[usize::from(*instr_ix)];
            if let InstructionValue::Function(fun) = &mut instr.value {
                merge_consecutive_blocks(env, &mut fun.lowered_function)?;
            }
        }

        // Can't merge value blocks and can't merge blocks with multiple
        // predecessors
        if block.kind != BlockKind::Block || block.predecessors.len() != 1 {
            return Ok(BlockRewriterAction::Keep(block));
        }

        let original_predecessor_id = block.predecessors.first().unwrap(); // length checked above
        let predecessor_id = merged.get(*original_predecessor_id);
        let predecessor = rewriter.block_mut(predecessor_id);
        if predecessor.kind != BlockKind::Block
            || !matches!(predecessor.terminal.value, TerminalValue::Goto(_))
        {
            // Can't merge value blocks, and we can't merge if the predecessor
            // has multiple successors (and isn't guaranteed to transfer here)
            return Ok(BlockRewriterAction::Keep(block));
        }

        // Replace phis in the merged block with canonical assignments to the single
        // operand value
        for phi in block.phis.iter_mut() {
            invariant(phi.operands.len() == 1, || {
                Diagnostic::invariant(ExpectedSingleOperandPhis { block: block_id }, None)
            })?;
            let (_, operand) = phi.operands.first().unwrap();
            // load the operand
            let temporary = env.new_temporary();
            let load = Instruction {
                id: predecessor.terminal.id,
                lvalue: IdentifierOperand {
                    identifier: temporary.clone(),
                    effect: None,
                },
                value: InstructionValue::LoadLocal(LoadLocal {
                    place: IdentifierOperand {
                        effect: None,
                        identifier: operand.clone(),
                    },
                }),
            };
            let load_ix = InstrIx::new(instructions.len() as u32);
            instructions.push(load);
            predecessor.instructions.push(load_ix);
            // store it into the phi id
            let store = Instruction {
                id: predecessor.terminal.id,
                lvalue: IdentifierOperand {
                    identifier: env.new_temporary(),
                    effect: None,
                },
                value: InstructionValue::StoreLocal(StoreLocal {
                    lvalue: LValue {
                        kind: InstructionKind::Reassign,
                        identifier: IdentifierOperand {
                            identifier: phi.identifier.clone(),
                            effect: None,
                        },
                    },
                    value: IdentifierOperand {
                        identifier: temporary,
                        effect: None,
                    },
                }),
            };
            let store_ix = InstrIx::new(instructions.len() as u32);
            instructions.push(store);
            predecessor.instructions.push(store_ix);
        }
        let BasicBlock {
            instructions,
            terminal,
            ..
        } = *block;
        predecessor.instructions.extend(instructions);
        predecessor.terminal = terminal;
        merged.merge(block_id, predecessor_id);

        has_changes = true;
        Ok(BlockRewriterAction::Remove)
    })?;

    if has_changes {
        mark_instruction_ids(&mut fun.body)?;
        mark_predecessors(&mut fun.body);
    }

    Ok(())
}

#[derive(Default)]
struct MergedBlocks {
    merged: HashMap<BlockId, BlockId>,
}

impl MergedBlocks {
    fn merge(&mut self, block: BlockId, into: BlockId) {
        let target = self.get(into);
        self.merged.insert(block, target);
    }

    fn get(&self, block: BlockId) -> BlockId {
        let mut current = block;
        while let Some(mapped) = self.merged.get(&current) {
            current = *mapped;
        }
        current
    }
}

#[derive(Debug, Error)]
#[error("Expected predecessor {predecessor} to exist")]
#[allow(dead_code)]
pub struct ExpectedPredecessorToExist {
    predecessor: BlockId,
}

#[derive(Debug, Error)]
#[error(
    "Expected block {block} with single predecessor to have no phis or
    phis with a single operand, found multiple operands"
)]
pub struct ExpectedSingleOperandPhis {
    block: BlockId,
}
